go/host/storage/hostdb/search.go

```diff
@@ -0,0 +1,139 @@
+package hostdb
+
+import (
+	"math/big"
+	"strconv"
+	"strings"
+
+	gethcommon "github.com/ethereum/go-ethereum/common"
+	"github.com/ten-protocol/go-ten/go/common"
+)
+
+// todo (@will) add pagination if results are large
+func Search(db HostDB, query string) (*common.SearchResponse, error) {
+	inputType := identifyInputType(query)
+
+	var results []*common.SearchResult
+
+	switch inputType {
+	case "hash":
+		// could be rollup, batch, or transaction hash
+		results = append(results, searchByHash(db, query)...)
+	case "number":
+		// could be batch height or sequence
+		results = append(results, searchByNumber(db, query)...)
+	default:
+		// try all searches
+		results = append(results, searchByHash(db, query)...)
+		results = append(results, searchByNumber(db, query)...)
+	}
+
+	searchResults := make([]common.SearchResult, len(results))
+	for i, result := range results {
+		searchResults[i] = *result
+	}
+
+	return &common.SearchResponse{
+		ResultsData: searchResults,
+		Total:       uint64(len(results)),
+	}, nil
+}
+
+func identifyInputType(query string) string {
+	query = strings.TrimPrefix(query, "0x")
+
+	if len(query) == 64 {
+		return "hash"
+	}
+
+	if len(query) == 40 {
+		return "address"
+	}
+
+	if _, err := strconv.ParseInt(query, 10, 64); err == nil {
+		return "number"
+	}
+
+	return "unknown"
+}
+
+func searchByHash(db HostDB, hash string) []*common.SearchResult {
+	var results []*common.SearchResult
+
+	trimmedHash := strings.TrimPrefix(hash, "0x")
+
+	rollup, err := GetRollupByHash(db, gethcommon.HexToHash(hash))
+	if err == nil {
+		results = append(results, &common.SearchResult{
+			Type:      "rollup",
+			Hash:      trimmedHash,
+			Timestamp: rollup.Timestamp,
+			ExtraData: map[string]interface{}{
+				"rollup": rollup,
+			},
+		})
+	}
```

Comment:
somethign like this everywhere:


```golang
 } else if !errros(err, notfound){
logger.Error("error getting...)
}
```

Reply:
I forgot to commit my stashed changes, my b. not sure how the transaction.go one got in but not the others